### PR TITLE
initial push

### PR DESCRIPTION
**Pull request by alexzhou-chata**

### CODE REVIEW

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js [hunk 1]
```diff
@@ -0,0 +1,214 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import './JoinColumnSelectionTable.scss'
+import { Select } from '../../Select'
+import { Icon } from '../../Icon'
+import { isListQuery, isAggregation, getGroupableColumns, getStringColumnIndices } from 'autoql-fe-utils'
+export default class JoinColumnSelectionTable extends React.Component {
```
Comment (eleon-chata-ai) [inline src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:7]: need to run prettier formatter on file

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+  constructor(props) {
+    super(props)
+    const selectableJoinColumnsLength = this.getSelectableJoinColumnsLength()
```
Comment (eleon-chata-ai) [inline src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:10]: can this be removed and used on line 13?

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    firstQueryResult: PropTypes.shape({}),
+    secondQueryResult: PropTypes.shape({}),
+    firstQueryFirstValue: PropTypes.string,
+    firstQuerySecondValue: PropTypes.string,
+    secondQueryFirstValue: PropTypes.string,
+    secondQuerySecondValue: PropTypes.string,
+    onFirstQueryFirstValueChange: PropTypes.func,
+    onFirstQuerySecondValueChange: PropTypes.func,
+    onSecondQueryFirstValueChange: PropTypes.func,
+    onSecondQuerySecondValueChange: PropTypes.func,
```
Comment (Shiv-Chata) [inline src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:38]: Does you need prop types for these values (line 31-38) if they are just state values?

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    onRemoveSecondValues: PropTypes.func,
+  }
+
+  static defaultProps = {
+    columnHeaders: [],
+    rowHeaders: [],
+    firstQueryOptions: [],
+    secondQueryOptions: [],
```
Comment (Shiv-Chata) [inline src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:46]: Is this unused?

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    onFirstQuerySecondValueChange: PropTypes.func,
+    onSecondQueryFirstValueChange: PropTypes.func,
+    onSecondQuerySecondValueChange: PropTypes.func,
+    onRemoveSecondValues: PropTypes.func,
+  }
+
+  static defaultProps = {
+    columnHeaders: [],
+    rowHeaders: [],
+    firstQueryOptions: [],
```
Comment (Shiv-Chata) [inline src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:45]: Is this unused?

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+    if (this.state.secondQuerySecondValue) {
+      this.props.onSecondQuerySecondValueChange(this.state.secondQuerySecondValue)
+      this.setState({ showSecondRow: true })
+    }
+  }
+
+  getDefaultJoinColumn = (queryResult) => {
+    const { defaultJoinColumn } = this.getDefaultJoinColumnAndDisplayNameAndJoinColumnsIndices(queryResult)
+    return defaultJoinColumn[0]
```
Comment (Shiv-Chata) [inline src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:71]: I would either add a ?.[0] or surround this with a try and catch

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  getSelectableJoinColumnsLength() {
```
Comment (Shiv-Chata) [inline src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:74]: I would wrap this in a try and catch

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+    return Math.min(firstSelectableColumns.length, secondSelectableColumns.length)
+  }
+  getQuerySelectableJoinColumns = (queryColumns) => {
+    return queryColumns
+      .map((col, index) => (getStringColumnIndices(queryColumns)?.stringColumnIndices?.includes(index) ? col : null))
+      .filter(Boolean)
+  }
+  getDefaultJoinColumnAndDisplayNameAndJoinColumnsIndices = (queryResult) => {
+    const columns = queryResult?.data?.data?.columns || []
```
Comment (Shiv-Chata) [inline src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:88]: I would this function in a try and catch

File: src/components/Notifications/RuleSimple/RuleSimple.js [hunk 9]
```diff
@@ -223,12 +216,129 @@ export default class RuleSimple extends React.Component {
...
+    const secondQuerySelectableJoinColumns = this.getQuerySelectableJoinColumns(secondQueryColumns)
+    return (
+      <div className='join-column-selection-prompt'>
+        <h4>Select join columns for your queries:</h4>
+        <p>Please choose the columns to join your first and second queries.</p>
+        <JoinColumnSelectionTable
+          columnHeaders={['Query 1', 'Query 2']}
+          rowHeaders={['1st Join', '2nd Join']}
+          firstQueryFirstOptions={firstQuerySelectableJoinColumns.map((col) => ({
+            value: col.name,
```
Comment (Shiv-Chata) [inline src/components/Notifications/RuleSimple/RuleSimple.js:281]: Would be nice to add ?.name

File: src/components/Notifications/RuleSimple/RuleSimple.js [hunk 10]
```diff
@@ -223,12 +216,129 @@ export default class RuleSimple extends React.Component {
...
     if (this.state.secondQueryResponse && this.state.secondQueryResponse !== prevState.secondQueryResponse) {
       setTimeout(() => {
         this.secondFieldSelectionGridRef?.current?.scrollIntoView({ behavior: 'smooth', block: 'end' })
       }, 100)
     }
+    if (this.state.secondQueryResult !== prevState.secondQueryResult && this.state.secondTermType === QUERY_TERM_TYPE) {
+      this.updateSecondFieldSelectionGrid()
+    }
+    if (this.state.secondQueryResult !== prevState.secondQueryResult && this.state.storedInitialData !== undefined) {
+      this.setState((prevState) => ({
```
Comment (Shiv-Chata) [inline src/components/Notifications/RuleSimple/RuleSimple.js:230]: Could this logic be moved to a function to increase readability? A function like handleResetSecondQueryJoinColumns

File: src/components/Notifications/RuleSimple/RuleSimple.js [hunk 11]
```diff
@@ -238,6 +348,36 @@ export default class RuleSimple extends React.Component {
     }
   }
 
+  getDefaultJoinColumnAndDisplayNameAndJoinColumnsIndices = (queryResult) => {
```
Comment (Shiv-Chata) [inline src/components/Notifications/RuleSimple/RuleSimple.js:351]: Is this not the same function in JoinColimnSelectionTable? If so it should be moved to utils function

File: src/components/Notifications/RuleSimple/RuleSimple.js [hunk 12]
```diff
@@ -238,6 +348,36 @@ export default class RuleSimple extends React.Component {
...
+      if (groupableColumns.length > 0) {
+        defaultJoinColumn = [columns[groupableColumns[0]].name]
+        defaultDisplayName = columns[groupableColumns[0]].display_name
+        defaultJoinColumnsIndices = [selectableJoinColumns.findIndex((col) => col.name === defaultJoinColumn[0])]
+      }
+    }
+
+    return { defaultJoinColumn, defaultDisplayName, defaultJoinColumnsIndices }
+  }
+  getQuerySelectableJoinColumns = (queryColumns) => {
```
Comment (Shiv-Chata) [inline src/components/Notifications/RuleSimple/RuleSimple.js:375]: Is this used outside of this file? This seem like a useful Utils function

File: src/components/Notifications/RuleSimple/RuleSimple.js [hunk 13]
```diff
@@ -333,31 +473,114 @@ export default class RuleSimple extends React.Component {
...
+
+    if (selectedNumberColumnName) {
+      compareColumnIndex = response?.data?.data?.columns?.findIndex((col) => col.name === selectedNumberColumnName)
+    } else {
+      compareColumnIndex = response?.data?.data?.columns?.findIndex((col) => isColumnNumberType(col) && col.is_visible)
+    }
+
+    return compareColumnIndex === -1 ? undefined : compareColumnIndex
+  }
+  getQuery = () => {
```
Comment (Shiv-Chata) [inline src/components/Notifications/RuleSimple/RuleSimple.js:495]: Wrap this in a try and catch

File: src/components/Notifications/RuleSimple/RuleSimple.js [hunk 14]
```diff
@@ -333,31 +473,114 @@ export default class RuleSimple extends React.Component {
...
+      ? Promise.resolve(this.props.queryResponse)
+      : runQueryOnly({
+          query: this.props.initialData?.[0]?.term_value,
+          ...getAuthentication(this.props.authentication),
+          ...getAutoQLConfig(this.props.autoQLConfig),
+          source: 'data_alert_first_query',
+          pageSize: 2,
+          allowSuggestions: false,
+        })
+    const fetchSecondQuery =
```
Comment (Shiv-Chata) [inline src/components/Notifications/RuleSimple/RuleSimple.js:507]: This complexity is a little tough to read. Is there a way to simplify this?

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js (hunk 1 above)
Comment (alexzhou-chata) [reply src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:7]: I just checked, it's already formatted by prettier

**Review by Shiv-Chata — COMMENTED** [review]
Looks good Alex! Well done. 

Left a couple small comments, and some notes. 

When going through the comments and PR again i would like you to think what should be included in the file and what could be moved to a utils file? Because this react-autoql package has very long files which make it hard to read. Also, when functions have a bunch of logic it is good to prepare them for errors, so if an error does occur the component can recover and move on

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js (hunk 2 above)
Comment (alexzhou-chata) [reply src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:10]: Yep sure

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js (hunk 3 above)
Comment (alexzhou-chata) [reply src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:38]: I do need those prop types as I do have 
this.props.firstQueryFirstValue and so on. At least it won't hurt anything?
<JoinColumnSelectionTable
          columnHeaders={['Query 1', 'Query 2']}
          rowHeaders={['1st Join', '2nd Join']}
          firstQueryFirstOptions={firstQuerySelectableJoinColumns.map((col) => ({
            value: col.name,
            label: col.display_name,
          }))}
          firstQuerySecondOptions={firstQuerySelectableJoinColumns.map((col) => ({
            value: col.name,
            label: col.display_name,
          }))}
          secondQueryFirstOptions={secondQuerySelectableJoinColumns.map((col) => ({
            value: col.name,
            label: col.display_name,
          }))}
          secondQuerySecondOptions={secondQuerySelectableJoinColumns.map((col) => ({
            value: col.name,
            label: col.display_name,
          }))}
          firstQueryResult={this.state.firstQueryResult}
          secondQueryResult={this.state.secondQueryResult}
          firstQueryFirstValue={this.state.firstQueryFirstValue}
          firstQuerySecondValue={this.state.firstQuerySecondValue}
          secondQueryFirstValue={this.state.secondQueryFirstValue}
          secondQuerySecondValue={this.state.secondQuerySecondValue}
          onFirstQueryFirstValueChange={this.handleFirstQueryFirstValueChange}
          onFirstQuerySecondValueChange={this.handleFirstQuerySecondValueChange}
          onSecondQueryFirstValueChange={this.handleSecondQueryFirstValueChange}
          onSecondQuerySecondValueChange={this.handleSecondQuerySecondValueChange}
          onRemoveSecondValues={this.handleRemoveSecondValues}
          isLoadingSecondQuery={this.state.isLoadingSecondQuery}
          storedInitialData={this.state.storedInitialData}
        />

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js (hunk 4 above)
Comment (alexzhou-chata) [reply src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:46]: Thx, I will remove it

File: src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js (hunk 5 above)
Comment (alexzhou-chata) [reply src/components/Notifications/JoinColumnSelectionTable/JoinColumnSelectionTable.js:45]: Thx, I will remove them

**Comment by alexzhou-chata** [comment]
> Looks good Alex! Well done.
> 
> Left a couple small comments, and some notes.
> 
> When going through the comments and PR again i would like you to think what should be included in the file and what could be moved to a utils file? Because this react-autoql package has very long files which make it hard to read. Also, when functions have a bunch of logic it is good to prepare them for errors, so if an error does occur the component can recover and move on

Thank you!

File: src/components/Notifications/RuleSimple/RuleSimple.js (hunk 11 above)
Comment (alexzhou-chata) [reply src/components/Notifications/RuleSimple/RuleSimple.js:351]: Sure,  you are right, it's better to move it to Utility librarys

File: src/components/Notifications/RuleSimple/RuleSimple.js (hunk 14 above)
Comment (alexzhou-chata) [reply src/components/Notifications/RuleSimple/RuleSimple.js:507]: I added a shouldFetchSecondQuery to make it more readable

**Review by Shiv-Chata — DISMISSED** [review]

**Review by Shiv-Chata — APPROVED** [review]

**Pull request merged**